### ISSUE TITLE
Problems with promotion of mixed-types

### ISSUE THREAD

**Issue by lbenet** [JuliaDiff/TaylorSeries.jl#94]
As reported in #92, arithmetic operations are subtle when dealing with `Taylor1{TaylorN{T}}`. The simplest case I've come to, which yields to a segmentation fault is:
```julia
julia> using TaylorSeries

julia> x = 1.0 + TaylorN(Float64,1,order=5)
 1.0 + 1.0 x₁ + 𝒪(‖x‖⁶)

julia> y = Taylor1(x, 16)
  1.0 + 1.0 x₁ + 𝒪(‖x‖⁶) + 𝒪(t¹⁷)

julia> x + y
```

The problem seems to be related with promotion.

**Comment by dpsanders** [comment]
What is the result of that operation expected to be? What if the `TaylorN` had two variables? Should this even be defined?

**Comment by lbenet** [comment]
To make myself clearer, reconsider what happens above, but in the following form:

```julia
julia> using TaylorSeries

julia> δx, δy = set_variables("δx δy")
2-element Array{TaylorSeries.TaylorN{Float64},1}:
  1.0 δx + 𝒪(‖x‖⁷)
  1.0 δy + 𝒪(‖x‖⁷)

julia> x = 1+Taylor1(δx,10)
  1.0 + 1.0 δx + 𝒪(‖x‖⁷) + 𝒪(t¹¹)

```

So, in a way, `x` is time dependent (`Taylor1`), and `δx` represent possible variations close to 1. (There is also a δy, so the `TaylorN` part is actually a 2-variable polynomial.) So altogether, it is a polynomial in 3 variables. So my example would be to consider `x + δx`, which should result in `1.0 + 2.0 δx + 𝒪(‖x‖⁷) + 𝒪(t¹¹)`.

The idea of having this type of things is that you integrate the time dependence, and are able to say things about small deviations close to the reference trajectory.

**Comment by PerezHz** [comment]
Using the latest commit of `fuse_operations` branch I noticed there seems to be still some trouble with `TaylorN{Taylor1{Float64}}`:

```julia
julia> using TaylorSeries

julia> p = set_variables("ξ", numvars=4, order=1);

julia> x1N = Taylor1([2+TaylorN(Float64,1,order=1),1+TaylorN(Float64,2,order=1)],10)
  2.0 + 1.0 ξ₁ + 𝒪(‖x‖²) + ( 1.0 + 1.0 ξ₂ + 𝒪(‖x‖²)) t + 𝒪(t¹¹)

julia> xN1 = convert( TaylorN{Taylor1{Float64}},x1N )
 ( 2.0 + 1.0 t + 𝒪(t¹¹)) + ( 1.0 + 𝒪(t¹¹)) ξ₁ + ( 1.0 t + 𝒪(t¹¹)) ξ₂ + 𝒪(‖x‖²)

julia> x1N^(3/2)
  2.82842712474619 + 2.1213203435596424 ξ₁ + 𝒪(‖x‖²) + ( 2.1213203435596424 + 0.5303300858899107 ξ₁ + 2.1213203435596424 ξ₂ + 𝒪(‖x‖²)) t + ( 0.2651650429449553 - 0.06629126073623881 ξ₁ + 0.5303300858899106 ξ₂ + 𝒪(‖x‖²)) t² + ( - 0.022097086912079608 + 0.016572815184059703 ξ₁ - 0.06629126073623882 ξ₂ + 𝒪(‖x‖²)) t³ + ( 0.0041432037960149265 - 0.005179004745018657 ξ₁ + 0.016572815184059706 ξ₂ + 𝒪(‖x‖²)) t⁴ + ( - 0.0010358009490037316 + 0.0018126516607565304 ξ₁ - 0.005179004745018659 ξ₂ + 𝒪(‖x‖²)) t⁵ + ( 0.00030210861012608837 - 0.0006797443727836988 ξ₁ + 0.0018126516607565306 ξ₂ + 𝒪(‖x‖²)) t⁶ + ( - 9.710633896909982e-5 + 0.0002670424321650245 ξ₁ - 0.0006797443727836989 ξ₂ + 𝒪(‖x‖²)) t⁷ + ( 3.3380304020628065e-5 - 0.00010848598806704122 ξ₁ + 0.0002670424321650246 ξ₂ + 𝒪(‖x‖²)) t⁸ + ( - 1.2053998674115689e-5 + 4.5202495027933845e-5 ξ₁ - 0.00010848598806704123 ξ₂ + 𝒪(‖x‖²)) t⁹ + ( 4.520249502793383e-6 - 1.9211060386871884e-5 ξ₁ + 4.520249502793385e-5 ξ₂ + 𝒪(‖x‖²)) t¹⁰ + 𝒪(t¹¹)

julia> xN1^(3/2)
ERROR: ArgumentError: Division does not define a Taylor1 polynomial
or its first non-zero coefficient is Inf/NaN.
Order k=10 => coeff[11]=NaN.
 in divfactorization(::TaylorSeries.Taylor1{Float64}, ::TaylorSeries.Taylor1{Float64}) at /Users/Jorge/.julia/v0.5/TaylorSeries/src/arithmetic.jl:479
 in /(::TaylorSeries.Taylor1{Float64}, ::TaylorSeries.Taylor1{Float64}) at /Users/Jorge/.julia/v0.5/TaylorSeries/src/arithmetic.jl:430
 in ./(::Array{TaylorSeries.Taylor1{Float64},1}, ::TaylorSeries.Taylor1{Float64}) at ./arraymath.jl:112
 in /(::TaylorSeries.HomogeneousPolynomial{TaylorSeries.Taylor1{Float64}}, ::TaylorSeries.Taylor1{Float64}) at /Users/Jorge/.julia/v0.5/TaylorSeries/src/arithmetic.jl:398
 in ^(::TaylorSeries.TaylorN{TaylorSeries.Taylor1{Float64}}, ::Float64) at /Users/Jorge/.julia/v0.5/TaylorSeries/src/power.jl:335
```

On the other hand, arithmetics with `Taylor1{TaylorN{Float64}}` seems to be working fine.

**Comment by lbenet** [comment]
I actually noticed something similar before you did 😄  !

The current last commit is https://github.com/JuliaDiff/TaylorSeries.jl/pull/95/commits/025cba8876571f4ec69b6be4b41e032e9ab0f81d .

**Comment by PerezHz** [comment]
Another conversion issue with `TaylorN{Taylor1{Float64}}`:

```julia
julia> using TaylorSeries

julia> p = set_variables("ξ", numvars=4, order=1)
4-element Array{TaylorSeries.TaylorN{Float64},1}:
  1.0 ξ₁ + 𝒪(‖x‖²)
  1.0 ξ₂ + 𝒪(‖x‖²)
  1.0 ξ₃ + 𝒪(‖x‖²)
  1.0 ξ₄ + 𝒪(‖x‖²)

julia> δx1 = TaylorN(Taylor1{Float64},1,order=1)
 ( 1.0 + 𝒪(t¹)) ξ₁ + 𝒪(‖x‖²)

julia> x = Taylor1(16)
 1.0 t + 𝒪(t¹⁷)

julia> x+δx1
 (  1.0 + 𝒪(t¹) t + 𝒪(t¹⁷)) + (  1.0 + 𝒪(t¹) + 𝒪(t¹)) ξ₁ + 𝒪(‖x‖²)

julia> typeof(ans)
TaylorSeries.TaylorN{TaylorSeries.Taylor1{TaylorSeries.Taylor1{Float64}}}

```

when adding a `Taylor1{Float64}` and a `TaylorN{Taylor1{Float64}}`, the result is a `TaylorN{Taylor1{Taylor1{{Float64}}}`, where a `TaylorN{Taylor1{Float64}}` is expected

**Comment by lbenet** [comment]
Thanks for reporting. You are right; I'll take a look on this.

**Comment by lbenet** [comment]
The new commit https://github.com/JuliaDiff/TaylorSeries.jl/pull/95/commits/e23dec7c5b1518b5355c1c354bf4ce2d7a26a8c3 should solve this.

**Comment by PerezHz** [comment]
Commit e23dec7 works great on local tests!

**Comment by lbenet** [comment]
Closing, since it is fixed in #95.
